Ginfo Integration

iv-mexx opened the pull request:
As discussed, here's the WIP pull request for my changes with the integration into the [Ginfo Map](https://conanexiles.ginfo.gg).

I've described in https://github.com/NullSoldier/serverthrall/pull/15 how to test the integrations.


As mentioned, I'm not really familiar with django, so I expect there's a lot that can be cleaned up, I've tried to not touch the existing code too much though.

Most of the work is done in the `ginfo` plugin which just takes a `character`. 
The characters position is transformed to our map coordinate system in 2 parts: 

* Linear x/y coordinates are transformed via a linear transformation with transformation parameters specific to our map
* Linear x'/y' coordinates are transformed to lat/lng coordinates via a [Spherical Mercator](https://en.wikipedia.org/wiki/Mercator_projection#The_spherical_model) transformation, which is used by Leaflet (the mapping framework we use). This is a general transformation not specific to our map.

Then, data for a marker is created and posted via the REST API to our firebase database.


If this is the first time posting, a new marker will be created via a POST request. 
It is then necessary to keep the UID of the created marker so that the same marker can later be updated via PATCH request. Otherwise we would be creating a new marker each time an littering the map with old markers...

iv-mexx committed:
Start integrating into ginfo map

iv-mexx committed:
Only post to ginfo, if ginfo_group_uid is given via parameter

iv-mexx committed:
Apply the correct transform so markers are now posted to Ginfo with the correct coordinates

iv-mexx committed:
Dynamically create markers and save the marker uid for updates

iv-mexx committed:
Pass on the ginfo access token for accesss control by firebase

iv-mexx committed:
Fix migration conflict after rebasing on master

iv-mexx committed:
Post markers with an existing creatorId

iv-mexx committed:
Add specific ginfo character model

iv-mexx committed:
Save ginfo_marker_uid in the ginfocharacter model, not in the general character model

iv-mexx committed:
Use requests built in json decoder

iv-mexx committed:
Remove wrong on_delete rule for GinfoCharacter

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,167 @@
+    """
+      GinfoPlugin provides methods to post a position to the ginfo firebase database
+    """
+    API_URL = 'https://ginfo-dev.firebaseio.com'
```
**note** change this before merge

iv-mexx commented:
@NullSoldier Please don't merge yet, we might have to change something because of security rules

iv-mexx committed:
Change layout of data thats pushed to firebase for better handling of access token.

This also removes the necessity to distinguish between POST and PATCH. Firebase UIDs are generated locally with the firebase algorithm and markers are always PATCHed

NullSoldier commented:
yeah, I'm waiting for you to take this out of WIP to finish reviewing / considering merge. I'll LGTM before we merge.

iv-mexx committed:
Target Ginfo production database

iv-mexx commented:
Alright, We've deployed the necessary update on our site, I've created another pull request for documentation ( https://github.com/NullSoldier/serverthrall/pull/20 ) and updated this PR.

I'll need to do a last quick test with the production system when I get home today, then this can be finished up! 😃

iv-mexx commented:
Alright, looks good from my side!

NullSoldier commented on api/migrations/0005_ginfocharacter.py:
```diff
@@ -0,0 +1,24 @@
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('ginfo_marker_uid', models.TextField()),
+                ('character', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='api.Character')),
```
You need to regenerate this migration, or just remove this on_delete.

iv-mexx replied:
I deleted and regenerated the migration, but `on_delete` was there again. I'll remove it by hand.

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,153 @@
+    API_URL = 'https://ginfo-34baa.firebaseio.com'
+    GAME = 'conanexiles'
+    MAP = 'conanexiles_desert'
+    GINFO_USER_UID = 'y67Efa3kddSuhb26UoUapIJakMh1'
```
Question: If these are created by GINFO user, does the admin still have the ability to delete the markers if they want? What is the implication of this user id being publically available? Do you need to make this the admins user id?

iv-mexx replied:
This is the UID of a User I've created specifically for ServerThrall, but its just an ordinary user. Having its UID public should be no problem. The only request possible without authentication is to post markers like ServerThrallAPI does, but you still need the access token for a group to do that. 
If you're an authenticated user, you could theoretically send group invites to this user. Other than that, theres nothing you can with the UID.

Right now we have it so that either everyone in a group can move/delete every marker, or at least admin can move/delete every marker (this is currently used only in special groups).

I was thinking about requiring the Users UID as a configuration parameter as well like the Group UID. But its nice to have a way for us to query for markers created by ServerThrall for statistics.

NullSoldier replied:
That sounds good to me. I was gonna recommend at least making this a user specifically for server thrall but I'm happy to see you've already done that.

iv-mexx replied:
Yeah I see, the name of the constant does not imply that its a specific user, I'll change that!

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,153 @@
+            @param group: UID of the ginfo group for which the marker should be posted
+            @param marker: UID of the marker that should be posted
+        """
+        return "games/" + self.GAME + "/groups/" + group + "/maps/" + self.MAP + '/markers/' + marker
```
Since strings are immutable in python, using format is a lot faster. It allocates a new string for every + operation.

```python
 return 'games/%s/groups/%s/maps/%s/markers/%s' % (self.GAME, group, self.MAP, marker)
```

iv-mexx replied:
👍

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,153 @@
+            @param group: UID of the ginfo group for which the marker should be posted
+            @param marker: UID of the marker that should be posted
+        """
+        return "groupAccessTokenVerifier/" + group + "/" + marker
```
Use format here too.

NullSoldier commented on api/plugins/ginfo/ginfo.py:
I'll upgrade this to use newrelic later

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,153 @@
+        ginfo_character, created = GinfoCharacter.objects.get_or_create(
+            character_id=character.id)
+        if created:
+            ginfo_character.ginfo_marker_uid = self.uid_generator.next_id()
```
Can this use pythons built in uuid class? I think that it satisfies all the needs that this class provides. If you use uuid v4 it takes into consideration the current time stamp, a seeded random number, and other factors. Collisions are also theoretically impossible.

```python
from uuid import uuid4
uuid4().hex
>>> '15160478f2154eb3ab55a0334c7cd7e2'
uuid4().hex
>>> '95f1fc01fa884641ab05f1d73b62f4b7'
```

NullSoldier commented on api/plugins/ginfo/ginfo.py:
```diff
@@ -0,0 +1,153 @@
+            character_id=character.id)
+        if created:
+            ginfo_character.ginfo_marker_uid = self.uid_generator.next_id()
+            ginfo_character.save()
```
Right now this ginfo charater can save, even if the ginfo request fails. if you change this so that it doesn't try to create the character until after the request fails, that would solve the problem. That way there aren't "fake" ginfo characters in the database that don't represent markers in ginfo, and you don't need a transaction.

```
ginfo_character = GinfoCharacter.objects.filter(character_id=character.id).first()

if ginfo_character is None:
    ginfo_character = GinfoCharacter(character_id=character.id, ginfo_marker_uid=self.uid_generator.next_id())

# requests.patch(...)

ginfo_character.save()
```

iv-mexx replied:
Is the `save` operation expensive when nothing has changed?

NullSoldier replied:
This approach as discussed is more of an eventually consistency model and I'm okay with the way the code is written. Revoking my comment.

NullSoldier commented on api/plugins/ginfo/ginfo.py:
this can crash if firebase returns no body, after line 117 you should maybe check for no body, or a 500 error and print a different error. I'll add newrelic error reporting later to handle this better.

iv-mexx replied:
I've wrapped decoding into try/except to also catch responses with a body that is not valid json

NullSoldier commented:
This looks pretty close, I don't like githubs new review system. It's pretty confusing but I gave it a try.

iv-mexx committed:
Regenerate migration and remove on_delete hook

iv-mexx committed:
Use string interpolation to build database paths

iv-mexx committed:
Rename UID constant

iv-mexx committed:
Wrap response json decoding into try/except to not crash at unexpected responses

iv-mexx committed:
Add source reference to firebase_pushid

iv-mexx committed:
Post to ginfo only when there is a change and wrap the call in a try/except

NullSoldier commented:
LGTM

NullSoldier commented:
Thank you for the contribution, I'll get this deployed soon.

iv-mexx commented:
Nice, thank you for your guidance and the reviews!  I'm excited to have this feature in ServerThrall and Ginfo! 😃
